### PR TITLE
Fix opsgenie plugin to use correct annotation to determine whether or not to create alerts

### PR DESCRIPTION
**Pull request by EdwardDowling**
Fixes Opsgenie annotation implementation and docs differing.

changelog: Breaking change for Opsgenie plugin users, role annotations must now contain "teleport.dev/notify-services" to receive notification on Opsgenie. "teleport.dev/schedules" is now the label used to determine auto approval flow

Verified that when 
- notify label and not approval label is set - alert is created
-  both set - alert is created and if the user is on call the alert is approved
-  just approval label - no alert created
- both set and user is not on-all - alert is created but not approved

Will need [3774](https://github.com/gravitational/teleport.e/pull/3774) merged after this to correctly instantiate the opsgneie app

### CODE REVIEW

File: docs/img/enterprise/plugins/opsgenie/add-requester-role.png
Comment (zmb3) [inline docs/img/enterprise/plugins/opsgenie/add-requester-role.png:1]: This is a large (1080p) image with small text.

What is it that you are trying to show by including this screenshot? If it's the content of the role, can we include that as text instead of a screenshot?

File: docs/pages/access-controls/access-request-plugins/opsgenie.mdx
Comment (zmb3) [inline docs/pages/access-controls/access-request-plugins/opsgenie.mdx:83]: Should we also include `teleport.dev/schedules` in the spec above? It's a bit confusing when the example shows only one annotation and the text after it discusses two annotations.

File: integrations/access/accessrequest/app.go
Comment (zmb3) [inline integrations/access/accessrequest/app.go:369]: If there was an error, doesn't that mean that `rec` is likely nil? (In which case this code would panic..)

File: integrations/access/opsgenie/bot.go
Comment (zmb3) [inline integrations/access/opsgenie/bot.go:64]:
```suggestion
	rawRecipients := make([]string, 0, len(recipients)
```

**Review by zmb3 — DISMISSED** [review]

File: docs/img/enterprise/plugins/opsgenie/add-requester-role.png
Comment (EdwardDowling) [reply docs/img/enterprise/plugins/opsgenie/add-requester-role.png:1]: It is included as text as well just bellow the screenshot, the purpose of the screenshot is really just to make it easier to see where in the web ui the role can be added from.

**Comment by ptgott** [comment]
@EdwardDowling I'm going to unsubscribe until we sort out the technical review. Let me know if you'd like a docs review!

File: docs/img/enterprise/plugins/opsgenie/add-requester-role.png
Comment (zmb3) [reply docs/img/enterprise/plugins/opsgenie/add-requester-role.png:1]: My vote would be to remove it.

File: integrations/access/accessrequest/app.go
Comment (r0mant) [inline integrations/access/accessrequest/app.go:363]: Didn't we agree that in master (v16) each annotation will be used for its intended purpose, and in backports we'll use "schedules" for both if it's the only one present? From our discussion:

```
- In v15/v14 to keep backwards compatibility the behavior should be:
  - When only [teleport.dev/schedules](http://teleport.dev/schedules) annotation is present, it is used for both notifications and auto-approval to keep the behavior there is now. This behavior should not be documented anywhere, this is just for compatibility because the mess we made.
  - When both notify-services and schedules are present, each is used for their own intended purpose.
- In master (v16), we keep notify-services and schedules for their own purposes without this side-effect of sending notifications. I guess this means you'll need to revert the revert of your PR again because that's what basically your master PR already did.
  - And add this to breaking changes in the changelog to tell users they must have "notify-services" defined to receive notifications in Opsgenie.
- Go through all Opsgenie documentation and RFDs and make sure all annotations everywhere are accurate and properly documented, including screenshots.
```

**Review by r0mant — CHANGES_REQUESTED** [review]
In addition to Zac's comments, I don't think this is implemented in the way we discussed.

File: docs/pages/access-controls/access-request-plugins/opsgenie.mdx
Comment (r0mant) [inline docs/pages/access-controls/access-request-plugins/opsgenie.mdx:80]:
```suggestion
The `teleport.dev/notify-services` annotation specifies the schedules the alert will be created for.
```

File: integrations/access/accessrequest/app.go
Comment (r0mant) [inline integrations/access/accessrequest/app.go:362]:
```suggestion
				log.Warningf("Failed to fetch Opsgenie recipient: %v", err)
```

**Review by r0mant — COMMENTED** [review]
The logic seems ok now, in addition to a few more suggestions I left can you also:

- Update the PR description with the test plan you went through when testing this fix.
- Update CHANGELOG.md to include an entry about this breaking change. If we don't include it there now, we will 100% forget before v16 release.

File: integrations/access/opsgenie/client.go
Comment (EdwardDowling) [reply integrations/access/opsgenie/client.go]: Have made the variables in the opsgenie plugin more consistent. The `defaultSchedules` that belongs to the opsgenie client is used in /e as well so to change it we might need to make some extra field fill both in then remove once /e is updated. 
The `Responders` field is called responder instead of schedules because in Opsgenie it can be either a user or a schedule and copied their naming convention to match the field in the API and on the web page for clarity. Would it make more sense to swap over to schedules since that is the only way we use them?

File: api/types/constants.go
Comment (r0mant) [inline api/types/constants.go:726]: Let's rename this one to `ReqAnnotationApproveSchedulesLabel` too then, for consistency.

**Review by r0mant — COMMENTED** [review]
@EdwardDowling This looks ok now, but before I approve I've a final request: please update the PR description with the test plan you performed testing this.

Also, when backporting, keep in mind that in backports we want to keep the backwards compatible behavior as we discussed:

```
In v15/v14 to keep backwards compatibility the behavior should be:
- When only teleport.dev/schedules annotation is present, it is used for both notifications and auto-approval to keep the behavior there is now. This behavior should not be documented anywhere, this is just for compatibility.
- When both notify-services and schedules are present, each is used for their own intended purpose.
```

**Comment by hugoShaka** [comment]
I added a commit implementing basic automatic approval tests. This commit also makes the approval user variable (like in the PagerDuty plugin). After merging this PR we'll need to merge this e PR: https://github.com/gravitational/teleport.e/pull/3737

**Review by hugoShaka — APPROVED** [review]
LGTM

@r0mant and @EdwardDowling please take a look at the additional tests I sent and see if you're comfortable with them

File: integrations/access/opsgenie/client.go
Comment (r0mant) [inline integrations/access/opsgenie/client.go:162]: So what happens if this fails for some reason?

I wonder if it would be more robust to move this logic to request resolution time. So we would store `RequestID` here, and then the plugin would attempt to query `AlertID` when it needs to resolve the incident. WDYT?

File: integrations/access/opsgenie/app.go
Comment (r0mant) [reply integrations/access/opsgenie/app.go:442]: Looking at your [other PR](https://github.com/gravitational/teleport.e/pull/3774/files#diff-7c919d60e8a9363ba673474c3a99a1fe393e36aa3374b43d76d9ae0c9c7c27ebR38), it looks like it's still set to `teleport.SystemAccessApproverUserName` there, right? What's the reason for replacing the constant with the config field here then, I don't think it's supposed to be configurable, or am I missing something?

Comment (EdwardDowling) [reply integrations/access/opsgenie/app.go:442]: Since I needed to change opsgenie.go in the other PR to correctly call instantiate the opsgenie plugin i swapped this to be configured there to match Pagerduty and jira.  I can swap it back to the constant instead if you think that makes more sense. This change was for consistency more than anything else.

Comment (r0mant) [reply integrations/access/opsgenie/app.go:442]: Got it, that's fine. Let's keep it.

File: integrations/access/opsgenie/client.go
Comment (EdwardDowling) [reply integrations/access/opsgenie/client.go:162]: Have been trying out implementing this this morning and I think it should work but it gets a bit messy now having to double check everywhere we interact with an alert whether or not we have the alert or the request id and resolving it if appropriate. I think we might end up having to change some of the shared plugin code as well if we go this route to allow the accessrequest/MessageData stuff to represent a 'message' that we dont have the ID of yet

Comment (EdwardDowling) [reply integrations/access/opsgenie/client.go:162]: In the event this fails it could go one of two ways depending on if the request ever does end up completing successfully after the retries fail. 
- 1: The alert is never created and that is why the retries failed - in this case the plugin just never interacts with that access request again.
- 2: The alert is eventually created after the retries all fail  - in this case the alert is created but the plugin wont be able to send updates or resolve it

Comment (r0mant) [reply integrations/access/opsgenie/client.go:162]: The auto-approval will still work in both of these cases, right? As long as we're able to determine if requester is on-call. Cause we never interact with alert/request ID to actually approve Teleport's request - just the alert will not be closed in Opsgenie, correct?

Comment (EdwardDowling) [reply integrations/access/opsgenie/client.go:162]: Yes, the notify error in opsgenie/app.go and the error from trying to resolve the request are put into an aggregate error instead of returning early.

Comment (r0mant) [reply integrations/access/opsgenie/client.go:162]: Ok, in that case, let's do this:

- Add a comment here explaining that if this fails, Teleport request approval and auto-approval will still work but incident in Opsgenie won't be auto-closed.
- Increase the timeout to something more than 30s, maybe a couple minutes?

**Review by r0mant — APPROVED** [review]
lgtm after you address https://github.com/gravitational/teleport/pull/38435#discussion_r1539602042

File: integrations/access/opsgenie/client.go
Comment (EdwardDowling) [reply integrations/access/opsgenie/client.go:166]: Added in the comment and upped the timeout, does this look okay now?

Comment (r0mant) [reply integrations/access/opsgenie/client.go:166]: Yeah I already approved so let's get this merged and do the backports.

**Pull request merged**
